python/equistore-operations/equistore/operations/sort.py

```diff
@@ -0,0 +1,183 @@
+    sort_block_dims_values: bool = True, # sort each block's values' according to the labels
+    descending: bool = False,
+    stable: bool = False
+    # numpy
+    # using kind='quicksort' if False and kind='stable' if True
+) -> TensorMap:
+    """
+    >>> import numpy as np
+    >>> import equistore
+    >>> from equistore import TensorBlock, TensorMap, Labels
+    >>> np.random.seed(1)
+    >>> block = TensorBlock(
+    ...     values=np.arange(9).reshape(3,3),
+    ...     samples=Labels(["structures", "centers"], np.array([[0,3], [0,1], [0,2]])),
+    ...     components=[],
+    ...     properties=Labels(["n", "l"], np.array([[2,0], [3,0], [1,0]])),
+    ... )
+    >>> keys = Labels(names=["key"], values=np.array([[0]]))
+    >>> print(block)
+    TensorBlock
+        samples (3): ['structures', 'centers']
+        components (): []
+        properties (3): ['n', 'l']
+        gradients: None
+    >>> block_sorted = equistore.sort_block(block, sort_block_dims_values=False)
+    >>> block_sorted # order within a label entry has been sorted lexicographical
+    TensorBlock
+        samples (3): ['centers', 'structures']
+        components (): []
+        properties (3): ['l', 'n']
+        gradients: None
+    >>> block_sorted.samples
+    Labels([(3, 0), (1, 0), (2, 0)],
+           dtype=[('centers', '<i4'), ('structures', '<i4')])
+    >>> block.values
+    array([[0, 1, 2],
+           [3, 4, 5],
+           [6, 7, 8]])
+    >>> block_sorted.values # but not the values
+    array([[0, 1, 2],
+           [3, 4, 5],
+           [6, 7, 8]])
+    >>> block_sorted_stepwise = equistore.sort_block(block, block_dims=["properties"])
+    >>> block_sorted_stepwise.values # properties are are sorted which are columns in this case
+    array([[2, 0, 1],
+           [5, 3, 4],
+           [8, 6, 7]])
+    >>> block_sorted_stepwise = equistore.sort_block(block_sorted_stepwise , block_dims=["samples"])
+    >>> block_sorted_stepwise.values # samples are are sorted which are rows in this case
+    array([[5, 3, 4],
+           [8, 6, 7],
+           [2, 0, 1]])
+    >>> np.all(equistore.sort_block(block).values == block_sorted_stepwise.values) # doing everything at once and verify correctness
+    True
+    """
+
+    blocks = []
+    for block in tensor.blocks():
+        blocks.append(
+            sort_block(
+                block=block, block_dims=block_dims, sort_metadata=sort_metadata, descending=descending, stable=stable
+            )
+        )
+    return TensorMap(tensor.keys, blocks)
+
+
+def sort_block(
+    block: TensorBlock,
+    block_dims: List[str] = None,
+    sort_block_dims_names: bool = True, # sort each block's labels' names lexicographical
+    sort_block_dims_values: bool = True, # sort each block's values' according to the labels
+    descending: bool = False,
+    stable: bool = False
+) -> TensorBlock:
+    """
+    TODO
+    """
+    if block_dims is None:
+        block_dims = ["samples", "components", "properties"]
```

Comment:
We should validate that `block_dims` only contains allowed values

Reply:
Also, I don't see the point of sorting `components`, since each component can only contain a single column (and I would be fine enforcing that this column is sorted in equistore-core, I can see no reason for it not to be).

If this is about re-ordering components, I would be against since components carry more meaning (in particular, gradients and values components are linked)

Reply:
In this code each component is sorted separately (within the component). So the components order is not changed. I also intuitively would not do it, since this is outside of the scope of the Labels object.

Reply:
>  I also intuitively would not do it, since this is outside of the scope of the Labels object.

What's `it` here? Sorting them, or enforcing that they are sorted? We already enforce some limitation on the Labels used for components (the limitations are checked when constructing a TensorBlock, not when constructing a Labels), so we could easily add more limitations.

https://github.com/lab-cosmo/equistore/blob/b9bdcf5b433b17746a8a12fc3a0ba65295d5a55c/equistore-core/src/blocks.rs#L87-L104

Reply:
I wouldn't sort them in this function in the sense changing the order between components (first component remains first component).

I dont have an opinion about enforcing sorting on components in equistore-core. Seems to me a bit arbitrary without any context.